### PR TITLE
CLN: Assorted cleanups

### PR DESCRIPTION
**Pull request by jbrockmendel**
Many of these are broken off from #30035.

### CODE REVIEW

File: pandas/core/dtypes/cast.py [hunk 1]
```diff
@@ -337,6 +337,21 @@ def changeit():
 
 
 def maybe_promote(dtype, fill_value=np.nan):
+    """
+    Find the minimal dtype that can hold both the given dtype and fill_value.
+
+    Parameters
+    ----------
+    dtype : np.dtype or ExceptionDtype
```
Comment (WillAyd) [inline pandas/core/dtypes/cast.py]: ExtensionDtype?

File: pandas/core/strings.py [hunk 2]
```diff
@@ -74,10 +74,12 @@ def cat_core(list_of_columns: List, sep: str):
     """
     if sep == "":
         # no need to interleave sep if it is empty
-        return np.sum(list_of_columns, axis=0)
+        arr_of_cols = np.asarray(list_of_columns, dtype=object)
```
Comment (WillAyd) [inline pandas/core/strings.py]: What is this cleaning up?

File: pandas/io/pytables.py [hunk 3]
```diff
@@ -3314,12 +3312,12 @@ def data_orientation(self):
     def queryables(self) -> Dict[str, Any]:
         """ return a dict of the kinds allowable columns for this object """
 
+        # mypy doesnt recognize DataFrame._AXIS_NAMES, so we re-write it here
```
Comment (WillAyd) [inline pandas/io/pytables.py]: What error was this throwing? Surprised it wouldn't see this

File: pandas/core/dtypes/cast.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/cast.py]: yep, will change

File: pandas/core/strings.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/strings.py]: this is for when numpy re-institutes the make-object-explicit thing from #30035.

File: pandas/io/pytables.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/io/pytables.py]: I think _AXIS_NAMES is defined dynamically in a classmethod.

File: pandas/core/strings.py (hunk 2 above)
Comment (WillAyd) [reply pandas/core/strings.py]: I think should hold off on these changes then until that comes back

Comment (jbrockmendel) [reply pandas/core/strings.py]: they reverted it specifically to give downstream time to update without breaking CIs

File: pandas/core/dtypes/cast.py [hunk 4]
```diff
@@ -337,6 +337,21 @@ def changeit():
 
 
 def maybe_promote(dtype, fill_value=np.nan):
+    """
```
Comment (jreback) [inline pandas/core/dtypes/cast.py]: should type at some point

File: pandas/core/dtypes/cast.py [hunk 5]
```diff
@@ -686,10 +700,12 @@ def maybe_upcast(values, fill_value=np.nan, dtype=None, copy: bool = False):
 
     Parameters
     ----------
-    values : the ndarray that we want to maybe upcast
+    values : ndarray or ExtensionArray
```
Comment (jreback) [inline pandas/core/dtypes/cast.py]: ideally type these at some point

File: pandas/core/strings.py (hunk 2 above)
Comment (jreback) [reply pandas/core/strings.py]: great this is good

File: pandas/io/pytables.py (hunk 3 above)
Comment (jreback) [reply pandas/io/pytables.py]: yes, it likely needs typing, i would rather do that and remove this from here (followup ok)

**Review by jreback — APPROVED** [review]
lgtm. small followup requests.

**Pull request merged**